File: tools/server_observer/recording_registry.py
import json
import threading
from pathlib import Path
from typing import Dict
from typing import Optional

from tools.recorder.recorder_logger import get_logger

logger = get_logger()

class RecordingRegistry:
    """Persisted registry of active and finished recordings to allow recovery."""

    def __init__(self, registry_path: Path):
        self.path = Path(registry_path)
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self._lock = threading.Lock()
        self.state: Dict[str, Dict[str, dict]] = {"recording": {}, "completed": {}, "failed": {}}
        self._load()

    def _load(self):
        if not self.path.exists():
            return
        try:
            with self.path.open("r") as f:
                data = json.load(f)
                if isinstance(data, dict):
                    self.state.update({k: v for k, v in data.items() if k in self.state})
        except Exception as exc:
            logger.warning(f"Could not read registry at {self.path}: {exc}")

    def _save(self):
        with self.path.open("w") as f:
            json.dump(self.state, f, indent=2)

    def mark_recording(self, game_id: int, scenario_id: int, replay_path: Optional[str]):
        with self._lock:
            self.state["recording"][str(game_id)] = {
                "scenario_id": scenario_id,
                "replay_path": replay_path,
            }
            self._save()

    def mark_completed(self, game_id: int):
        with self._lock:
            meta = self.state["recording"].pop(str(game_id), None)
            if meta is None:
                meta = {}
            self.state["completed"][str(game_id)] = meta
            self._save()

    def mark_failed(self, game_id: int, reason: str = ""):
        with self._lock:
            meta = self.state["recording"].pop(str(game_id), None)
            if meta is None:
                meta = {}
            meta["reason"] = reason
            self.state["failed"][str(game_id)] = meta
            self._save()

    def is_known(self, game_id: int) -> bool:
        sid = str(game_id)
        return any(sid in bucket for bucket in self.state.values())

    def active(self) -> Dict[int, dict]:
        return {int(k): v for k, v in self.state["recording"].items()}

File: tools/server_observer/src/recording_registry.cpp
#include "recording_registry.hpp"
#include <fstream>
#include <filesystem>
#include <iostream>

namespace fs = std::filesystem;

RecordingRegistry::RecordingRegistry(const std::string& registry_path)
    : path_(registry_path)
{
    // Create parent directories if they don't exist
    fs::path path(path_);
    if (path.has_parent_path()) {
        fs::create_directories(path.parent_path());
    }
    
    // Initialize state with default buckets
    state_ = {
        {"recording", json::object()},
        {"completed", json::object()},
        {"failed", json::object()}
    };
    
    load();
}

void RecordingRegistry::load() {
    std::ifstream file(path_);
    if (!file.is_open()) {
        return;  // File doesn't exist yet
    }
    
    try {
        json data;
        file >> data;
        
        if (data.is_object()) {
            for (auto& [key, value] : data.items()) {
                if (state_.contains(key)) {
                    state_[key] = value;
                }
            }
        }
    } catch (const std::exception& e) {
        std::cerr << "Warning: Could not read registry at " << path_ << ": " << e.what() << std::endl;
    }
}

void RecordingRegistry::save() {
    std::ofstream file(path_);
    if (!file.is_open()) {
        std::cerr << "Error: Could not write registry to " << path_ << std::endl;
        return;
    }
    
    file << state_.dump(2);
}

void RecordingRegistry::mark_recording(int game_id, int scenario_id, const std::string& replay_path) {
    std::lock_guard<std::mutex> lock(lock_);
    
    json meta = {
        {"scenario_id", scenario_id}
    };
    
    if (!replay_path.empty()) {
        meta["replay_path"] = replay_path;
    }
    
    state_["recording"][std::to_string(game_id)] = meta;
    save();
}

void RecordingRegistry::mark_completed(int game_id) {
    std::lock_guard<std::mutex> lock(lock_);
    
    std::string game_id_str = std::to_string(game_id);
    json meta;
    
    if (state_["recording"].contains(game_id_str)) {
        meta = state_["recording"][game_id_str];
        state_["recording"].erase(game_id_str);
    }
    
    state_["completed"][game_id_str] = meta;
    save();
}

void RecordingRegistry::mark_failed(int game_id, const std::string& reason) {
    std::lock_guard<std::mutex> lock(lock_);
    
    std::string game_id_str = std::to_string(game_id);
    json meta;
    
    if (state_["recording"].contains(game_id_str)) {
        meta = state_["recording"][game_id_str];
        state_["recording"].erase(game_id_str);
    }
    
    if (!reason.empty()) {
        meta["reason"] = reason;
    }
    
    state_["failed"][game_id_str] = meta;
    save();
}

bool RecordingRegistry::is_known(int game_id) const {
    std::lock_guard<std::mutex> lock(lock_);
    std::string game_id_str = std::to_string(game_id);
    
    for (const auto& [bucket_name, bucket] : state_.items()) {
        if (bucket.is_object() && bucket.contains(game_id_str)) {
            return true;
        }
    }
    
    return false;
}

std::map<int, json> RecordingRegistry::active() const {
    std::lock_guard<std::mutex> lock(lock_);
    std::map<int, json> result;
    
    if (state_.contains("recording") && state_["recording"].is_object()) {
        for (const auto& [key, value] : state_["recording"].items()) {
            try {
                int game_id = std::stoi(key);
                result[game_id] = value;
            } catch (const std::exception&) {
                // Skip invalid game IDs
            }
        }
    }
    
    return result;
}
